components/support/shutdown/src/lib.rs

```diff
@@ -0,0 +1,85 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+License, v. 2.0. If a copy of the MPL was not distributed with this
+* file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use parking_lot::Mutex;
+use std::fmt;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Weak;
+
+static IN_SHUTDOWN: AtomicBool = AtomicBool::new(false);
+lazy_static::lazy_static! {
+   static ref REGISTERED_INTERRUPTS: Mutex<Vec<Weak<dyn ShutdownInterrupt>>> = Mutex::new(Vec::new());
+}
+
+/// Initiate shutdown:
+///
+/// - All registered `ShutdownInterrupt` instances will have their `interrupt()` methods called.
+/// - After this `err_if_shutdown` will always return a ShutdownError
+pub fn shutdown() {
+    IN_SHUTDOWN.store(true, Ordering::Relaxed);
+    for weak in REGISTERED_INTERRUPTS.lock().iter() {
+        if let Some(interrupt) = weak.upgrade() {
+            interrupt.interrupt()
+        }
+    }
+}
+
+/// Restart after a shutdown
```

Comment:
How is this used? I wonder if we can soften expectations here, saying something like "used only for testing" or similar? Or even better, only make it available in the test cfg? Specifically, I'm thinking that to make this truly reliable we'd probably need additional protections, ensuring other threads aren't currently interrupting etc?